crates/red_knot_python_semantic/docs/kind_of_types.md

```diff
@@ -0,0 +1,201 @@
+- A "literal module type": a type representing a single known module.
+
+Since it is known that all inhabitants of a given singleton type share the same memory address,
+singleton types in unions can be safely narrowed by identity,
+using [the operators `is` and `is not`](https://snarky.ca/unravelling-is-and-is-not/):
+
+```py
+def f(x: str | None):
+    if x is None:
+        ...  # x can be narrowed to `str`
+    else:
+        ...  # x can be narrowed to `None`
+```
+
+All singleton types are also sealed types, closed types and single-value types
+(see below for definitions).
+
+## Sealed types
+
+A sealed type is a type for which it is known
+that there is only a finite and pre-determined set of inhabitants at runtime.
+It follows from this that for every sealed type,
+there is also only a finite and pre-determined set of *subtypes* of that type.
+
+A sealed type in Python has the property
+that its proper subtypes[^2] are all disjunct from one another,
+and the union of all the proper subtypes of a sealed type is exactly equal to the sealed type.
+
+All singleton types are sealed types; however, not all sealed types are singleton types.
+
+Examples of sealed types (other than the singleton types listed above) are:
+
+- `bool`:
+
+    - The only inhabitants of `bool` at runtime are the constants `True` and `False`.
+    - The only proper subtypes of `bool` are `Literal[True]`, `Literal[False]`, and `Never`.
+        `Literal[True]`, `Literal[False]` and `Never` are all disjunct types,
+        and their union is exactly equal to `bool`.
+
+- Enums: consider the following enum class:
+
+    ```py
+    from enum import Enum, auto
+
+    class Foo(Enum):
+        X = auto()
+        Y = auto()
+        Z = auto()
+    ```
+
+    - The only inhabitants of the `Foo` type at runtime are the enum `Foo`'s members:
+        `Foo.X` and `Foo.Y`
+    - The only proper subtypes of `Foo` are `Literal[Foo.X]`, `Literal[Foo.Y]`, `Literal[Foo.Z]`, and `Never`.
+        `Literal[Foo.X]`, `Literal[Foo.Y]`, `Literal[Foo.Z]` and `Never` are all disjunct,
+        and the union `Literal[Foo.X, Foo.Y, Foo.Z] | Never` is exactly equal to the type `Foo`.
+
+Because a sealed type is equivalent to the union of all of its proper subtypes,
+for any given sealed type `X` where the only proper subtypes of `X` are `A`, `B` and `C`,
+`X & ~A == B | C`. To give a Python example:
+
+```py
+from enum import Enum, auto
+
+class X(Enum):
+    A = auto()
+    B = auto()
+    C = auto()
+
+
+def f(var: X):
+    ...  # the type of `var` here is `X` (which is equivalent to `Literal[X.A, X.B, X.C]`)
+
+    if var is X.A:
+        ...  # var is narrowed to `Literal[X.A]`
+    else:
+        ...  # var is narrowed to `Literal[X.B, X.C]`
+```
+
+## Single-value types
+
+A single-value type is a non-empty type for which it is known
+that all inhabitants of the type are equivalent with respect to their runtime value.
```

Comment:
It might be considered implied by "equivalence", but I think we should be explicit from the beginning that we are talking here about equivalence with respect to Python equality as defined by `__eq__` and `__ne__` implementations. "Equivalent with respect to their runtime value" is less clear what exactly we are talking about.

Reply:
The introductory paragraph at the top of the document states:

> This document provides definitions for various kinds of types, with examples of Python types that satisfy these definitions.

I'm _trying_ to present generalised definitions in the first 1-2 sentences of each section, that would apply outside of the specific context of Python, then in later sentences expand on the initial definition and link it back to the concept of Python types. So I would _prefer_ not to mention `__eq__` or `__ne__` in the first sentence here, as this is very specific to the way equivalence between arbitrary objects is generally implemented in Python.

Reply:
I didn't realize you were aiming to initially present a definition of each type abstracted from the Python context. I'm not sure how concretely valuable that is, since this is documentation for red-knot, which is a Python type-checker. And I think it's a bit challenging for "single-valued type" specifically, because what we mean by the term in red-knot _is_ specific to the relation defined by `__eq__` and `__ne__`. It's certainly possible, though.

Reply:
I think the core issue here is that the phrase "equivalent with respect to their runtime value" doesn't have any clear meaning in an abstract type system sense. So if you want to present an abstract definition, you'll have to start by saying "Assuming some equivalence relation is defined on type inhabitants..." and go from there.

(You can probably also skip discussion of reflexivity, transitivity, and symmetry, and relegate that to a footnoted link to e.g. https://en.wikipedia.org/wiki/Equivalence_relation, if you are using the term "equivalence relation", since an equivalence relation is defined to require all of those properties.)

Reply:
> I didn't realize you were aiming to initially present a definition of each type abstracted from the Python context. I'm not sure how concretely valuable that is, since this is documentation for red-knot, which is a Python type-checker.

It's not documentation for _users_ of red-knot though; it's documentation for red-knot developers to help them (us) think about type-system concepts. I personally find it helpful to consider these concepts in a more abstract way initially, before then linking the concepts back to the specific context of Python :-)